### PR TITLE
[TTS] Unify Read/Write Patterns for Regular and NonSweepable Entries in TS

### PR DESCRIPTION
**Pull request by gmaretic**
General
Before this PR:
Transactions only persisted information about writes to tables that are being swept to the sweep queue.

After this PR:

==COMMIT_MSG==
AtlasDB now persists the start timestamp of every write transaction that only writes to tables that are not sweepable into the sweep queue.
==COMMIT_MSG==

Priority:
Soon

Concerns / possible downsides (what feedback would you like?):
Is there a better design?

Is documentation needed?:
Yes

Compatibility
Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?:
No*

Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?:
Yes, but only for writes and reads to rows that will only be accessed by the new version

The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.):
This has no correctness implications yet, so yes

Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?:
No

Does this PR need a schema migration?
No

Testing and Correctness
What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?:
N/A

What was existing testing like? What have you done to improve it?:
Added tests

If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.:
N/A

If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?:
N/A

Execution
How would I tell this PR works in production? (Metrics, logs, etc.):
We have a new metrics tag for this

Has the safety of all log arguments been decided correctly?:
N/A

Will this change significantly affect our spending on metrics or logs?:
Should be fine

How would I tell that this PR does not work in production? (monitors, etc.):
Metrics

If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?:
Rollback is fine

If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC):
N/A

Scale
Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.:
No

Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?:
No

Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?:
No

Development Process
Where should we start reviewing?:

If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?:
Was actually 2 PRs

Please tag any other people who should be aware of this PR:
@jeremyk-91
@sverma30

### CODE REVIEW

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/SweepableTimestampsTest.java [hunk 1]
```diff
@@ -189,15 +189,34 @@ public void getCorrectNextTimestampWhenMultipleCandidates() {
...
+    }
+
+    @Test
+    public void noClashWithNonSweepable() {
+        writeToDefaultCellCommitted(sweepableTimestamps, 250L, TABLE_CONS);
+        writeToDefaultCellUncommitted(sweepableTimestamps, 50_250L, TABLE_THOR);
+        writeToDefaultCellAborted(sweepableTimestamps, 100_250L, TABLE_NOTH);
+
+        assertThat(readConservative(shardCons)).hasValue(0L);
+        assertThat(sweepableTimestamps.nextTimestampPartition(ShardAndStrategy.nonSweepable(), 0L, 200_000L))
```
Comment (sudiksha27) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/SweepableTimestampsTest.java:206]: nit: we should extract a method for sake of uniformity

**Review by sudiksha27 — COMMENTED** [review]
🎖️ Thanks for the refactor. I think this makes a lot of sense + cleaner than our original approach. I do think we should add a few more tests. Broadly -
1. Reader works for no-op and sweep makes progress (I think first half of this test exists)
2. The queue is cleared for noon strategy after reading
3. No cleaning  no-op strategies
4. Unit tests around PartitionInfo, WriteInfo, BatchInfo
5. We should add some tests for ser/deser for SweepQueueTableRow indevelop and ensure it does not break in this branch.
6. Most importantly this should not interfere with existing sweep. Existing tests should suffice
Some of these already exist but we should still have some form of integration and ETE testing before rolling out.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepQueueDeleter.java [hunk 2]
```diff
@@ -54,6 +54,9 @@ public class SweepQueueDeleter {
      * sentinels or not.
      */
     public void sweep(Collection<WriteInfo> unfilteredWrites, Sweeper sweeper) {
+        if (sweeper == Sweeper.NO_OP) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepQueueDeleter.java:59]: that's a bit much but hopefully now it's better

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCells.java [hunk 3]
```diff
@@ -269,13 +273,17 @@ private TimestampsToSweep getTimestampsToSweepDescendingAndCleanupAborted(
...
             long commitTs = startToCommitTs.get(startTs);
             if (commitTs == TransactionConstants.FAILED_COMMIT_TS) {
                 lastSweptTs = startTs;
-                writesByStartTs.get(startTs).forEach(write -> cellsToDelete
-                        .computeIfAbsent(write.tableRef(), ignore -> HashMultimap.create())
-                        .put(write.cell(), write.timestamp()));
+                abortedTimestamps.add(startTs);
+                writesByStartTs.get(startTs).stream()
+                        .filter(info -> info.writeRef() != null)
+                        .forEach(write -> cellsToDelete
```
[the remark this answers is not in the document]
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCells.java:284]: No, I checked all usages, but I changed it to optional now -- not sure it's better :D

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableTimestamps.java [hunk 4]
```diff
@@ -60,13 +59,13 @@ Map<Cell, byte[]> populateCells(PartitionInfo info, List<WriteInfo> writes) {
 
     private SweepableTimestampsTable.SweepableTimestampsRow computeRow(PartitionInfo partitionInfo) {
         return SweepableTimestampsTable.SweepableTimestampsRow.of(
-                partitionInfo.shard(),
+                partitionInfo.shardAndStrategy().shard(),
                 SweepQueueUtils.tsPartitionCoarse(partitionInfo.timestamp()),
-                partitionInfo.isConservative().persistToBytes());
+                partitionInfo.shardAndStrategy().strategy().persistToBytes());
     }
```
[the remark this answers is not in the document]
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableTimestamps.java:65]: this is not a bad idea but the gradle task keeps failing and I can't

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/WriteInfoPartitioner.java [hunk 5]
```diff
@@ -55,52 +58,54 @@ public WriteInfoPartitioner(KeyValueService kvs, Supplier<Integer> numShards) {
...
-        return partitionWritesByShardStrategyTimestamp(filterOutUnsweepableTables(writes));
-    }
-
-    @VisibleForTesting
-    List<WriteInfo> filterOutUnsweepableTables(List<WriteInfo> writes) {
-        return writes.stream()
-                .filter(writeInfo -> getStrategy(writeInfo).isPresent())
-                .collect(Collectors.toList());
+        Preconditions.checkArgument(!writes.isEmpty(), "There must be at least one write.");
+        Preconditions.checkArgument(
```
[the remark this answers is not in the document]
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/WriteInfoPartitioner.java:67]: Better?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/clear/DefaultTableClearer.java [hunk 6]
```diff
@@ -62,18 +62,20 @@ public DefaultTableClearer(KeyValueService kvs, ImmutableTimestampSupplier immut
     }
 
     @Override
+    @SuppressWarnings("ConstantConditions") // no writeInfo that reaches here will have a null writeReference
     public Collection<WriteInfo> filter(Collection<WriteInfo> writeInfos) {
         Set<TableReference> tablesToCareAbout = getConservativeTables(
-                writeInfos.stream().map(WriteInfo::tableRef).collect(Collectors.toSet()));
+                writeInfos.stream().map(info -> info.writeRef().tableRef()).collect(Collectors.toSet()));
```
[the remark this answers is not in the document]
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/clear/DefaultTableClearer.java:69]: I added it to the interface java doc

**Review by sudiksha27 — APPROVED** [review]
lgtm!

**Pull request merged**